unitTests/magicbox/ResultPreviewsGridTest.ts

```diff
@@ -0,0 +1,456 @@
+                  expect(grid.moveUp()).toBeFalsy();
+                  expect(grid.moveLeft()).toBeFalsy();
+                  for (let currentRow = 0; currentRow < rows; currentRow += 1) {
+                    const isLastRow = currentRow === rows - 1;
+                    const columnsInCurrentRow = isLastRow ? columnsInLastRow : columns;
+                    let currentDirection = 1;
+                    for (
+                      let currentColumn = 0;
+                      currentColumn < columnsInCurrentRow && currentColumn >= 0;
+                      currentColumn += currentDirection
+                    ) {
+                      if (currentRow === 0) {
+                        expect(grid.moveUp()).toBeFalsy();
+                      }
+                      if (currentColumn === 0) {
+                        expect(grid.moveLeft()).toBeFalsy();
+                      }
+                      if (isLastRow || (currentRow === rows - 2 && currentColumn >= columnsInLastRow)) {
+                        expect(grid.moveDown()).toBeFalsy();
+                      }
+                      if (currentColumn === columnsInCurrentRow - 1) {
+                        expect(grid.moveRight()).toBeFalsy();
+                        currentDirection = -1;
+                      }
+                      expect(grid.getSelectedPreviewElement().outerHTML).toEqual(
+                        receivedPreviews[currentRow * columns + currentColumn].dom.outerHTML
+                      );
+                      if (currentDirection === 1 && currentColumn < columnsInCurrentRow - 1) {
+                        expect(grid.moveRight()).toBeTruthy(`can move right to (${currentColumn + currentDirection}, ${currentRow})`);
+                      } else if (currentDirection === -1 && currentColumn > 0) {
+                        expect(grid.moveLeft()).toBeTruthy(`can move left to (${currentColumn + currentDirection}, ${currentRow})`);
+                      }
+                    }
+                    if (!isLastRow) {
+                      expect(grid.moveDown()).toBeTruthy(`can move down to (0, ${currentRow + 1})`);
+                    }
+                  }
+                  expect(grid.moveFirst()).toBeTruthy('can move back to origin');
+                  expect(grid.moveRight()).toBeTruthy('can move right from origin again');
+                  expect(grid.selectKeyboardFocusedSelection().dom.outerHTML).toEqual(receivedPreviews[1].dom.outerHTML);
+                  dataset.previewCopiesInResolvedOrder.forEach((preview, i) => {
+                    if (i === 1) {
+                      expect(preview.onSelect).toHaveBeenCalledTimes(1);
+                    } else {
+                      expect(preview.onSelect).not.toHaveBeenCalled();
+                    }
+                  });
+                });
+                done();
+              });
+            });
+          });
+
+          describe('asynchronously', () => {
```

Comment:
I wanted to test cases where some of the parameters are promises rather than arrays. This is mostly for testing that elements aren't created until every promise is resolved and testing that promises can be interrupted cleanly by another call.